### PR TITLE
Adding more protections against invalid messages

### PR DESCRIPTION
**Pull request by khaines**
(Work in progress) 
Will address items related to https://github.com/weaveworks/cortex/issues/767
- Added a last ditch effort in the aws storage client to just effectively drop any batches that result in a ValidationException. This keeps them from being endlessly retried.

### CODE REVIEW

**Comment by tomwilkie** [comment]
I feel pretty confident that anyone sending label values >2k is doing it wrong, but just to check - does anyone this this should be a flag?

**Comment by bboreham** [comment]
There are some similar things controlled by flag - `ingester.max-series-per-user`, `ingester.max-series-per-metric`.  I don't mind particularly.

**Comment by cboggs** [comment]
Label values approaching 2k feels pretty wrong to me, but maybe there's a use case I'm not considering.

:+1: on adding it as a flag, though. I'm curious how label value length impacts indexing across the various storage backends...? Seems the flag would help vendors mitigate "bad" behavior depending on their backend.

**Comment by khaines** [comment]
Having this as a flag would certainly allow adaptation to different backends, or to allow someone running an installation to be more restrictive than the storage layer they are running against.

File: .gitignore [hunk 1]
```diff
@@ -13,3 +13,4 @@ pkg/ingester/client/cortex.pb.go
 pkg/ring/ring.pb.go
 bazel-*
 images/
+.idea/
```
Comment (bboreham) [inline .gitignore]: suspect this wasn't meant to be in the PR?

File: pkg/chunk/aws/storage_client.go [hunk 2]
```diff
@@ -248,6 +248,12 @@ func (a storageClient) BatchWrite(ctx context.Context, input chunk.WriteBatch) e
 				unprocessed.TakeReqs(requests, -1)
 				backoff.Wait()
 				continue
+			}else if ok && awsErr.Code() == "ValidationException"{
```
Comment (bboreham) [inline pkg/chunk/aws/storage_client.go]: Please run `go fmt`

Comment (bboreham) [inline pkg/chunk/aws/storage_client.go]: Can we get the desired effect without a hard-coded string?
Next-best would be a string constant.

File: pkg/ingester/ingester.go [hunk 3]
```diff
@@ -340,9 +357,9 @@ func (i *Ingester) append(ctx context.Context, sample *model.Sample) error {
 		return httpgrpc.Errorf(http.StatusBadRequest, "sample with timestamp %v is older than the maximum accepted age", sample.Timestamp)
 	}
 
-	if err := util.ValidateSample(sample); err != nil {
+	if err := ValidateSample(sample, i.cfg.MaxLengthLabelName, i.cfg.MaxLengthLabelValue); err != nil {
 		level.Error(util.WithContext(ctx, util.Logger)).Log("msg", "error validating sample", "err", err)
-		return nil
+		return err
```
Comment (bboreham) [inline pkg/ingester/ingester.go]: What difference does this change make?
Doesn't need to be in the code, but I would like the PR and/or commit message to say.

File: .gitignore (hunk 1 above)
Comment (khaines) [reply .gitignore]: No it was not... removed in latest push

File: pkg/chunk/aws/storage_client.go (hunk 2 above)
Comment (khaines) [reply pkg/chunk/aws/storage_client.go]: Corrected

Comment (khaines) [reply pkg/chunk/aws/storage_client.go]: Had to go with next best thing. The dynamo library doesn't have it's own exported constant for this error, like they do for `dynamodb.ErrCodeProvisionedThroughputExceededException`

File: pkg/ingester/ingester.go (hunk 3 above)
Comment (khaines) [reply pkg/ingester/ingester.go]: Thanks for calling this out. This is a change I intended to revert... simple edit but with potential large impact. The implication is that returning err here will tell callers of ingester.Push that the data is invalid, vs silently dropping it. I am unsure how all callers would react here, so this is not something that should survive the PR IMO

**Comment by khaines** [comment]
@bboreham got the latest updates in and took care of any feedback that needed updates, so it should be good to; pending any other feedback items.

File: .gitignore [hunk 4]
```diff
@@ -12,4 +12,4 @@ cmd/lite/lite
 pkg/ingester/client/cortex.pb.go
 pkg/ring/ring.pb.go
 bazel-*
-images/
+images/
```
Comment (tomwilkie) [inline .gitignore]: Y u no like newlines @khaines? :-)

File: pkg/chunk/aws/storage_client.go [hunk 5]
```diff
@@ -620,6 +627,12 @@ func (a storageClient) getDynamoDBChunks(ctx context.Context, chunks []chunk.Chu
 				unprocessed.TakeReqs(requests, -1)
 				backoff.Wait()
 				continue
+			} else if ok && awsErr.Code() == validationException {
+				// this read will never work, so the only option is to drop the offending request and continue.
+				// TODO: add more debug options for capturing data/telemetry about the offending items?
+				level.Warn(util.Logger).Log("Error while fetching data from Dynamo: %v", awsErr)
+				level.Debug(util.Logger).Log("Dropped request details: \n%v", requests)
+				continue
```
Comment (tomwilkie) [inline pkg/chunk/aws/storage_client.go]: Probably worth pulling these changes out into a separate PR; for instance, we should export a counter of dropped writes at very least, and potentially even drop the entire chunk write somehow.

File: pkg/ingester/ingester.go [hunk 6]
```diff
@@ -81,8 +86,17 @@ type Config struct {
...
+	MaxLengthLabelName  int
+	MaxLengthLabelValue int
+
+	// For testing, you can override the address and ID of this ingester
+	addr                  string
+	infName               string
+	id                    string
+	skipUnregister        bool
 	ingesterClientFactory func(addr string, cfg client.Config) (client.IngesterClient, error)
+	KVClient              ring.KVClient
```
Comment (tomwilkie) [inline pkg/ingester/ingester.go]: This smell like a bad merge or rebase to me, @khaine?

File: pkg/ingester/ingester.go [hunk 7]
```diff
@@ -107,6 +121,19 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
...
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		level.Error(util.Logger).Log("msg", "failed to get hostname", "err", err)
+		os.Exit(1)
+	}
+
+	f.StringVar(&cfg.infName, "ingester.interface", "eth0", "Name of network interface to read address from.")
+	f.StringVar(&cfg.addr, "ingester.addr", "", "IP address to register into consul.")
+	f.StringVar(&cfg.id, "ingester.id", hostname, "ID to register into consul.")
```
Comment (tomwilkie) [inline pkg/ingester/ingester.go]: Ditto bad merge / rebase?

File: pkg/ingester/validate.go [hunk 8]
```diff
@@ -21,7 +16,7 @@ const (
 )
 
 // ValidateSample returns an err if the sample is invalid
-func ValidateSample(s *model.Sample) error {
+func ValidateSample(s *model.Sample, maxLabelNameLength, maxLabelValueLength int) error {
```
Comment (tomwilkie) [inline pkg/ingester/validate.go]: Perhaps we should have a `ValidationConfig` that we embed in the `IngesterConfig`, as I'd also want to add the limits from #578 as a flag.

**Comment by tomwilkie** [comment]
@khaines added some comments; let me know if you want me to do this and get them in, I'd like to get this merge soon.

File: pkg/ingester/ingester.go (hunk 6 above)
Comment (khaines) [reply pkg/ingester/ingester.go]: certainly does... I'll try tidy this today

File: pkg/ingester/ingester.go (hunk 7 above)
Comment (khaines) [reply pkg/ingester/ingester.go]: certainly does... I'll try tidy this today

File: pkg/ingester/validate.go (hunk 8 above)
Comment (khaines) [reply pkg/ingester/validate.go]: moved these into their own config struct

File: pkg/chunk/aws/storage_client.go (hunk 5 above)
Comment (khaines) [reply pkg/chunk/aws/storage_client.go]: Adding a counter is certainly something I can add in a separate PR. As for dropping the chunk write, currently it drops the entire batch, which can be a bit heavy handed but it didn't seem at the time that the error data from dynamo provided any information to determine what element of the batch was the problem.

**Comment by tomwilkie** [comment]
LGTM!  I'll file a follow up for a counter from dropped items in the dynamodb backend.

**Comment by bboreham** [comment]
Failed lint https://circleci.com/gh/weaveworks/cortex/2186

**Comment by bboreham** [comment]
PR to fix lint #847

**Comment by khaines** [comment]
Thanks Bryan!

**Pull request merged**